clash-lib/src/Clash/Normalize/Transformations/DEC.hs

```diff
@@ -475,79 +452,59 @@ mkDisjointGroup
   -> NormalizeSession (Term,[Term])
 mkDisjointGroup inScope (fun,(seen,cs)) = do
     tcm <- Lens.view tcCache
-    let argss    = Foldable.toList cs
-        argssT   = zip [0..] (List.transpose argss)
-        (sharedT,distinctT) = List.partition (areShared tcm inScope . fmap (first stripTicks) . snd) argssT
-        -- TODO: find a better solution than "maybe undefined fst . uncons"
-        shared   = map (second (maybe (error "impossible") fst . List.uncons)) sharedT
-        distinct = map (Either.lefts) (List.transpose (map snd distinctT))
-        cs'      = fmap (zip [0..]) cs
-        cs''     = removeEmpty
-                 $ fmap (Either.lefts . map snd)
-                        (if null shared
-                           then cs'
-                           else fmap (filter (`notElem` shared)) cs')
-    (distinctCaseM,distinctProjections) <- case distinct of
-      -- only shared arguments: do nothing.
-      [] -> return (Nothing,[])
-      -- Create selectors and projections
-      (uc:_) -> do
-        let argTys = map (inferCoreTypeOf tcm) uc
-        disJointSelProj inScope argTys cs''
-    let newArgs = mkDJArgs 0 shared distinctProjections
-    case distinctCaseM of
-      Just lb -> return (Letrec [lb] (mkApps fun newArgs), seen)
-      Nothing -> return (mkApps fun newArgs, seen)
-
--- | Create a single selector for all the representable distinct arguments by
--- selecting between tuples. This selector is only ('Just') created when the
--- number of representable uncommmon arguments is larger than one, otherwise it
--- is not ('Nothing').
---
--- It also returns:
---
--- * For all the non-representable distinct arguments: a selector
--- * For all the representable distinct arguments: a projection out of the tuple
---   created by the larger selector. If this larger selector does not exist, a
---   single selector is created for the single representable distinct argument.
+    let argLen = case Foldable.toList cs of
+                   [] -> error ($curLoc <> "mkDisjointGroup: no disjoint groups")
+                   l:_ -> length l
+        csT :: [CaseTree (Either Term Type)]
+        csT = map (\i -> fmap (!!i) cs) [0..(argLen-1)]
+    (lbs,newArgs) <- List.mapAccumLM (\lbs c -> do
+                    let cL :: [Either Term Type]
```

Comment:
```suggestion
    (lbs,newArgs) <- List.mapAccumLM $ \lbs c -> do
      ...
```

or even

```suggestion
    (lbs,newArgs) <- 
      List.mapAccumLM $ \lbs c -> do
        ...
```

will make this function way less cramped.

Reply:
The `$` is not an option because `mapAccumLM` takes 3 arguments

Reply:
ahh sad